[BUG] MyPy test failures on main

Looks like mypy is failing on the current main. I am not sure what changed / how this got into main unless it was just caused by a update to mypy itself.

Example: this PR which changed nothing about the code https://github.com/microsoft/graspologic/pull/908
Test output (unit-and-doc-test): 
```{pytb}
Run mypy ./graspologic
graspologic\match\qap.py:420: error: Argument "key" to "sorted" has incompatible type "Callable[[Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]], Union[Sequence[Sequence[Sequence[Sequence[Any]]]], Any, _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]"; expected "Callable[[Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]], Union[SupportsDunderLT, SupportsDunderGT]]"
graspologic\match\qap.py:420: error: Value of type "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]" is not indexable
graspologic\match\qap.py:420: error: Incompatible return value type (got "Union[Sequence[Sequence[Sequence[Sequence[Any]]]], Any, _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]", expected "Union[SupportsDunderLT, SupportsDunderGT]")
graspologic\match\qap.py:433: error: Item "int" of "Union[int, RandomState, Generator, None]" has no attribute "permutation"
graspologic\match\qap.py:433: error: Item "None" of "Union[int, RandomState, Generator, None]" has no attribute "permutation"
graspologic\match\qap.py:457: error: Item "int" of "Union[int, RandomState, Generator, None]" has no attribute "uniform"
graspologic\match\qap.py:457: error: Item "None" of "Union[int, RandomState, Generator, None]" has no attribute "uniform"
graspologic\cluster\autogmm.py:559: error: Incompatible types in assignment (expression has type "List[Optional[RandomState]]", variable has type "ndarray[Any, dtype[signedinteger[Any]]]")
graspologic\embed\base.py:439: error: <nothing> has no attribute "ndim"
graspologic\embed\base.py:441: error: <nothing> has no attribute "ndim"
graspologic\embed\base.py:444: error: <nothing> has no attribute "shape"
graspologic\embed\base.py:446: error: <nothing> has no attribute "shape"
graspologic\embed\base.py:449: error: Incompatible types in assignment (expression has type "Union[ndarray[Any, Any], Any]", variable has type "List[Union[ndarray[Any, Any], Any]]")
graspologic\embed\case.py:205: error: Incompatible types in assignment (expression has type "floating[Any]", variable has type "float")
Found 14 errors in 4 files (checked 79 source files)
Error: Process completed with exit code 1.
```

Side note: when I get the latest version of mypy and run it locally on `main`, I actually get 
```
Found 393 errors in 55 files (checked 78 source files)
``` 

I wont post the whole traceback but these errors involve basically every module in the library. Many are private functions.

**A few larger points for discussion**: 
- Would anyone be opposed to taking mypy out of unit-and-doc-test and putting it as it's own check, much like black has its own code-format-check? It seems unnecessary to check on every os/version like we do for actual tests.
- Can someone explain to me what mypy gives us that beartype doesn't? I know it is static and beartype is dynamic, right? but in practice what is this doing for us? 
   - If nothing, can we remove? 
- What is our policy on mypy compatibility for legacy code that is private? 
- any tips on getting mypy to tell me what the function is that is actually causing the error? --verbose did not do it for me.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -93,7 +93,7 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Set up Python ${{matrix.python_version}} ${{matrix.os}}
-        uses: actions/setup-python@v1
+        uses: actions/setup-python@v2
         with:
           python-version: ${{matrix.python_version}}
       - name: Install dependencies Python ${{matrix.python_version}} ${{matrix.os}}

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -32,7 +32,7 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Set up Python 3.8
-        uses: actions/setup-python@v1
+        uses: actions/setup-python@v2
         with:
           python-version: 3.8
       - name: "Run Reference Documentation Generation"
@@ -60,7 +60,7 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Set up Python 3.8
-        uses: actions/setup-python@v1
+        uses: actions/setup-python@v2
         with:
           python-version: 3.8
       - name: Run Format Check
@@ -75,7 +75,7 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Set up Python 3.8
-        uses: actions/setup-python@v1
+        uses: actions/setup-python@v2
         with:
           python-version: 3.8
       - name: Run Test Coverage
@@ -113,7 +113,7 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: Set up Python 3.8
-        uses: actions/setup-python@v1
+        uses: actions/setup-python@v2
         with:
           python-version: 3.8
       - name: Install dependencies

--- a/graspologic/align/seedless_procrustes.py
+++ b/graspologic/align/seedless_procrustes.py
@@ -349,11 +349,11 @@ class SeedlessProcrustes(BaseAlign):
         m, _ = Y.shape
 
         if self.init == "2d":
-            P_matrices = np.zeros((2 ** d, n, m))
-            Q_matrices = np.zeros((2 ** d, d, d))
-            objectives = np.zeros(2 ** d)
+            P_matrices = np.zeros((2**d, n, m))
+            Q_matrices = np.zeros((2**d, d, d))
+            objectives = np.zeros(2**d)
             # try 2^d different initializations
-            for i in range(2 ** d):
+            for i in range(2**d):
                 initial_Q = _sign_flip_matrix_from_int(i, d)
                 P_matrices[i], Q_matrices[i] = P, Q = self._iterative_ot(
                     X, Y, initial_Q

--- a/graspologic/cluster/autogmm.py
+++ b/graspologic/cluster/autogmm.py
@@ -556,7 +556,7 @@ class AutoGMMCluster(BaseCluster):
                 np.iinfo(np.int32).max, size=len(processed_param_grid)
             )
         else:
-            seeds = [self.random_state] * len(processed_param_grid)
+            seeds = np.array([self.random_state] * len(processed_param_grid))
 
         n = X.shape[0]
         if self.max_agglom_size is None or n <= self.max_agglom_size:

--- a/graspologic/embed/base.py
+++ b/graspologic/embed/base.py
@@ -401,7 +401,7 @@ class BaseEmbedMulti(BaseSpectralEmbed):
         self.diag_aug = diag_aug
 
     def _check_input_graphs(
-        self, graphs: List[GraphRepresentation]
+        self, graphs: Union[List[GraphRepresentation], np.ndarray]
     ) -> Union[AdjacencyMatrix, List[AdjacencyMatrix]]:
         """
         Checks if all graphs in list have same shapes.
@@ -426,6 +426,8 @@ class BaseEmbedMulti(BaseSpectralEmbed):
             If all graphs do not have same shape, or input list is empty or has
             one element.
         """
+        out: Union[List[AdjacencyMatrix], np.ndarray]
+
         # Convert input to np.arrays
         # This check is needed because np.stack will always duplicate array in memory.
         if isinstance(graphs, (list, tuple)):

--- a/graspologic/embed/case.py
+++ b/graspologic/embed/case.py
@@ -202,7 +202,7 @@ class CovariateAssistedEmbed(BaseSpectralEmbed):
 
         # just use the ratio of the leading eigenvalues for the
         # tuning parameter, or the closest value in its possible range.
-        self.alpha_ = np.float_(L_top / YYt_top)
+        self.alpha_ = L_top / YYt_top
 
         return self
 

--- a/graspologic/embed/mds.py
+++ b/graspologic/embed/mds.py
@@ -197,7 +197,7 @@ class ClassicalMDS(BaseEstimator):
             dissimilarity_matrix = self._compute_euclidean_distances(X=X)
 
         J = _get_centering_matrix(dissimilarity_matrix.shape[0])
-        B = J @ (dissimilarity_matrix ** 2) @ J * -0.5
+        B = J @ (dissimilarity_matrix**2) @ J * -0.5
 
         n_components = self.n_components
 
@@ -216,7 +216,7 @@ class ClassicalMDS(BaseEstimator):
 
         self.n_components_ = len(D)
         self.components_ = U
-        self.singular_values_ = D ** 0.5
+        self.singular_values_ = D**0.5
         self.dissimilarity_matrix_ = dissimilarity_matrix
         self.n_features_in_ = X.shape[1]
 

--- a/graspologic/embed/n2v.py
+++ b/graspologic/embed/n2v.py
@@ -212,12 +212,12 @@ def _learn_embeddings(
     # Documentation - https://radimrehurek.com/gensim/models/word2vec.html
     model = Word2Vec(
         walks,
-        size=dimensions,
+        vector_size=dimensions,
         window=window_size,
         min_count=0,
         sg=1,  # Training algorithm: 1 for skip-gram; otherwise CBOW
         workers=workers,
-        iter=iterations,
+        epochs=iterations,
         seed=random_seed,
     )
 

--- a/graspologic/inference/latent_position_test.py
+++ b/graspologic/inference/latent_position_test.py
@@ -232,8 +232,8 @@ def _difference_norm(
             X1 = X1 / np.linalg.norm(X1, ord="fro")
             X2 = X2 / np.linalg.norm(X2, ord="fro")
         elif test_case == "diagonal-rotation":
-            normX1 = np.sum(X1 ** 2, axis=1)
-            normX2 = np.sum(X2 ** 2, axis=1)
+            normX1 = np.sum(X1**2, axis=1)
+            normX2 = np.sum(X2**2, axis=1)
             normX1[normX1 <= 1e-15] = 1
             normX2[normX2 <= 1e-15] = 1
             X1 = X1 / np.sqrt(normX1[:, None])

--- a/graspologic/layouts/nooverlap/_quad_node.py
+++ b/graspologic/layouts/nooverlap/_quad_node.py
@@ -900,7 +900,7 @@ class _QuadNode:
                 overlapping_node.color = "#F1FD00"  # Yellow
                 # print ("not None, a: %g, b: %g, c: %g" %(a, b, c), node_to_move.node_id, node_to_move.size, overlapping_node.size)
                 # print ("original(%g,%g), current(%g,%g), overlap(%g,%g)" %(node_to_move.original_x, node_to_move.original_y, new_x, new_y, overlapping_node.x, overlapping_node.y))
-                angle_c = math.acos((a ** 2 + b ** 2 - c ** 2) / (2 * a * b))
+                angle_c = math.acos((a**2 + b**2 - c**2) / (2 * a * b))
                 len_c_new = node_to_move.size + overlapping_node.size + _EPSILON
                 angle_a_new = math.asin(a * math.sin(angle_c) / len_c_new)
                 angle_b_new = 180 - math.degrees(angle_c) - math.degrees(angle_a_new)
@@ -909,10 +909,10 @@ class _QuadNode:
                 )
                 # print ("slope: %g, angle c: %g, new angle a: %g, newlenC: %g, new angle a: %g, new lenB %g" %(slope_ca, math.degrees(angle_c), math.degrees(angle_a_new), len_c_new, math.degrees(angle_a_new), new_len_b))
                 x_new_plus = node_to_move.original_x + math.sqrt(
-                    new_len_b ** 2 / (1 + slope_ca ** 2)
+                    new_len_b**2 / (1 + slope_ca**2)
                 )
                 x_new_neg = node_to_move.original_x - math.sqrt(
-                    new_len_b ** 2 / (1 + slope_ca ** 2)
+                    new_len_b**2 / (1 + slope_ca**2)
                 )
                 x_plus_diff = x_new_plus - new_x
                 x_neg_diff = x_new_neg - new_x
@@ -1006,7 +1006,7 @@ class _QuadNode:
                 denominator = 2 * a * b
                 if 0 == denominator:
                     denominator = 0.00000001
-                value = (a ** 2 + b ** 2 - c ** 2) / denominator
+                value = (a**2 + b**2 - c**2) / denominator
                 if value >= 1:
                     value = 0.999999
                 elif value <= -1:
@@ -1020,10 +1020,10 @@ class _QuadNode:
                 )
                 # print ("slope: %g, angle c: %g, new angle a: %g, newlenC: %g, new angle a: %g, new lenB %g" %(slope_ca, math.degrees(angle_c), math.degrees(angle_a_new), len_c_new, math.degrees(angle_a_new), new_len_b))
                 x_new_plus = node_to_move.original_x + math.sqrt(
-                    new_len_b ** 2 / (1 + slope_ca ** 2)
+                    new_len_b**2 / (1 + slope_ca**2)
                 )
                 x_new_neg = node_to_move.original_x - math.sqrt(
-                    new_len_b ** 2 / (1 + slope_ca ** 2)
+                    new_len_b**2 / (1 + slope_ca**2)
                 )
                 x_plus_diff = x_new_plus - new_x
                 x_neg_diff = x_new_neg - new_x

--- a/graspologic/match/qap.py
+++ b/graspologic/match/qap.py
@@ -3,11 +3,11 @@
 # original code can be found here
 # https://github.com/scipy/scipy/blob/master/scipy/optimize/_qap.py
 
+import numbers
 import operator
 from typing import Any, Optional, Union
 
 import numpy as np
-from scipy._lib._util import check_random_state
 from scipy.optimize import OptimizeResult, linear_sum_assignment
 from typing_extensions import Literal
 
@@ -416,9 +416,8 @@ def _quadratic_assignment_faq(
     # check outlier cases
     if n == 0 or partial_match_value.shape[0] == n:
         # Cannot assume partial_match is sorted.
-        partial_match_value = np.row_stack(
-            sorted(partial_match_value, key=lambda x: x[0])
-        )
+        sort_inds = np.argsort(partial_match_value[:, 0])
+        partial_match_value = partial_match_value[sort_inds]
         score = _calc_score(A, B, s_value, partial_match_value[:, 1])
         res = {"col_ind": partial_match_value[:, 1], "fun": score, "nit": 0}
         return OptimizeResult(res)
@@ -571,3 +570,37 @@ def _doubly_stochastic(P: np.ndarray, tol: float = 1e-3) -> np.ndarray:
         P_eps = r[:, None] * P * c
 
     return P_eps
+
+
+# copy-pasted from scipy scipy._lib._util
+# which was copy-pasted from scikit-learn utils/validation.py
+# this was just modified to add proper typing for returns
+# also, shouldn't have been importing private function from scipy anyway
+def check_random_state(
+    seed: Union[None, int, np.random.RandomState, np.random.Generator]
+) -> Union[np.random.RandomState, np.random.Generator]:
+    """Turn seed into a np.random.RandomState instance
+
+    If seed is None (or np.random), return the RandomState singleton used
+    by np.random.
+    If seed is an int, return a new RandomState instance seeded with seed.
+    If seed is already a RandomState instance, return it.
+    If seed is a new-style np.random.Generator, return it.
+    Otherwise raise ValueError.
+    """
+    if seed is None or seed is np.random:
+        return np.random.mtrand._rand
+    if isinstance(seed, (numbers.Integral, np.integer)):
+        int_seed: int = int(seed)  # necessary for typing/mypy
+        return np.random.RandomState(int_seed)
+    if isinstance(seed, np.random.RandomState):
+        return seed
+    try:
+        # Generator is only available in numpy >= 1.17
+        if isinstance(seed, np.random.Generator):
+            return seed
+    except AttributeError:
+        pass
+    raise ValueError(
+        "%r cannot be used to seed a numpy.random.RandomState" " instance" % seed
+    )

--- a/graspologic/models/sbm_estimators.py
+++ b/graspologic/models/sbm_estimators.py
@@ -226,7 +226,7 @@ class SBMEstimator(BaseGraphEstimator):
         n_blocks: int = self.block_p_.shape[0]
         n_parameters = 0
         if self.directed:
-            n_parameters += n_blocks ** 2
+            n_parameters += n_blocks**2
         else:
             n_parameters += int(n_blocks * (n_blocks + 1) / 2)
         if hasattr(self, "vertex_assignments_"):
@@ -445,7 +445,7 @@ class DCSBMEstimator(BaseGraphEstimator):
         n_blocks: int = self.block_p_.shape[0]
         n_parameters = 0
         if self.directed:
-            n_parameters += n_blocks ** 2  # B matrix
+            n_parameters += n_blocks**2  # B matrix
         else:
             n_parameters += int(n_blocks * (n_blocks + 1) / 2)  # Undirected B matrix
         if hasattr(self, "vertex_assignments_"):

--- a/graspologic/pipeline/embed/adjacency_spectral_embedding.py
+++ b/graspologic/pipeline/embed/adjacency_spectral_embedding.py
@@ -154,7 +154,7 @@ def adjacency_spectral_embedding(
     check_argument(svd_solver_iterations >= 1, "svd_solver_iterations must be positive")
 
     check_argument(
-        svd_seed is None or 0 <= svd_seed <= 2 ** 32 - 1,
+        svd_seed is None or 0 <= svd_seed <= 2**32 - 1,
         "svd_seed must be a nonnegative, 32-bit integer",
     )
 

--- a/graspologic/pipeline/embed/laplacian_spectral_embedding.py
+++ b/graspologic/pipeline/embed/laplacian_spectral_embedding.py
@@ -165,7 +165,7 @@ def laplacian_spectral_embedding(
     check_argument(svd_solver_iterations >= 1, "svd_solver_iterations must be positive")
 
     check_argument(
-        svd_seed is None or 0 <= svd_seed <= 2 ** 32 - 1,
+        svd_seed is None or 0 <= svd_seed <= 2**32 - 1,
         "svd_seed must be a nonnegative, 32-bit integer",
     )
 

--- a/graspologic/pipeline/embed/omnibus_embedding.py
+++ b/graspologic/pipeline/embed/omnibus_embedding.py
@@ -150,7 +150,7 @@ def omnibus_embedding_pairwise(
     check_argument(svd_solver_iterations >= 1, "svd_solver_iterations must be positive")
 
     check_argument(
-        svd_seed is None or 0 <= svd_seed <= 2 ** 32 - 1,
+        svd_seed is None or 0 <= svd_seed <= 2**32 - 1,
         "svd_seed must be a nonnegative, 32-bit integer",
     )
 

--- a/graspologic/simulations/simulations.py
+++ b/graspologic/simulations/simulations.py
@@ -277,7 +277,7 @@ def er_nm(
     # compute max number of edges to sample
     if loops:
         if directed:
-            max_edges = n ** 2
+            max_edges = n**2
             msg = "n^2"
         else:
             max_edges = n * (n + 1) // 2

--- a/setup.cfg
+++ b/setup.cfg
@@ -29,7 +29,7 @@ include_package_data = True
 install_requires =
     anytree>=2.8.0
     beartype>=0.9.0
-    gensim>=3.8.0,<=3.9.0  # methods signatures changed in the 4.0.0beta release
+    gensim>=4.0.0
     graspologic-native>=1.1.1
     hyppo>=0.2.2
     joblib>=0.17.0  # Older versions of joblib cause issue #806.  Transitive dependency of hyppo.

--- a/setup.cfg
+++ b/setup.cfg
@@ -35,7 +35,7 @@ install_requires =
     joblib>=0.17.0  # Older versions of joblib cause issue #806.  Transitive dependency of hyppo.
     matplotlib>=3.0.0,!=3.3.*
     networkx>=2.1
-    numpy>=1.8.1,<1.22
+    numpy>=1.8.1
     POT>=0.7.0
     seaborn>= 0.11.0
     scikit-learn>=0.22.0

--- a/setup.cfg
+++ b/setup.cfg
@@ -35,7 +35,7 @@ install_requires =
     joblib>=0.17.0  # Older versions of joblib cause issue #806.  Transitive dependency of hyppo.
     matplotlib>=3.0.0,!=3.3.*
     networkx>=2.1
-    numpy>=1.8.1
+    numpy>=1.8.1,<1.22
     POT>=0.7.0
     seaborn>= 0.11.0
     scikit-learn>=0.22.0

--- a/tests/test_models.py
+++ b/tests/test_models.py
@@ -420,15 +420,15 @@ class TestDCSBM(unittest.TestCase):
         labels = self.labels
         e = DCSBMEstimator(directed=True)
         e.fit(graph)
-        assert e._n_parameters() == (n_verts + n_class - 1 + n_class ** 2)
+        assert e._n_parameters() == (n_verts + n_class - 1 + n_class**2)
 
         e = DCSBMEstimator(directed=True)
         e.fit(graph, y=labels)
-        assert e._n_parameters() == (n_verts + n_class ** 2)
+        assert e._n_parameters() == (n_verts + n_class**2)
 
         e = DCSBMEstimator(directed=True, degree_directed=True)
         e.fit(graph, y=labels)
-        assert e._n_parameters() == (2 * n_verts + n_class ** 2)
+        assert e._n_parameters() == (2 * n_verts + n_class**2)
 
         e = DCSBMEstimator(directed=False)
         e.fit(graph, y=labels)
@@ -594,4 +594,4 @@ def hardy_weinberg(theta):
     """
     Maps a value from [0, 1] to the hardy weinberg curve.
     """
-    return np.array([theta ** 2, 2 * theta * (1 - theta), (1 - theta) ** 2]).T
+    return np.array([theta**2, 2 * theta * (1 - theta), (1 - theta) ** 2]).T

--- a/tests/test_select_dimension.py
+++ b/tests/test_select_dimension.py
@@ -16,7 +16,7 @@ def generate_data(n=10, elbows=3, seed=1):
     Generate data matrix with a specific number of elbows on scree plot
     """
     np.random.seed(seed)
-    x = np.random.binomial(1, 0.6, (n ** 2)).reshape(n, n)
+    x = np.random.binomial(1, 0.6, (n**2)).reshape(n, n)
     xorth = orth(x)
     d = np.zeros(xorth.shape[0])
     for i in range(0, len(d), int(len(d) / (elbows + 1))):

